protostar/git.py

```diff
@@ -0,0 +1,215 @@
+import subprocess
+import re
+import os.path
+from typing import Dict, Optional
+from pathlib import Path
+from functools import wraps
+
+from protostar.utils.package_info import PackageInfo
+from protostar.protostar_exception import ProtostarException
+
+# pylint: disable=subprocess-run-check
+
+# set to true when debugging
+GIT_VERBOSE = False
+SHARED_KWARGS = (
+    {"check": True}
+    if GIT_VERBOSE
+    else {
+        "check": True,
+        "stdout": subprocess.DEVNULL,
+        "stderr": subprocess.DEVNULL,
+    }
+)
+
+CREDENTIALS = [
+    "-c",
+    'user.name="Protostar"',
+    "-c",
+    'user.email="protostar@protostar.protostar"',
+]
+
+
+class ProtostarGitException(ProtostarException):
+    def __init__(self, message: str, details: Optional[str] = None):
+        super().__init__("Error while executing Git command:\n" + message, details)
+
+
+class NotARepositoryException(ProtostarGitException):
+    pass
+
+
+def wrap_git_exception(func):
+    @wraps(func)
+    def wrapper(*args, **kwargs):
+        try:
+            func(*args, **kwargs)
+        except subprocess.CalledProcessError as ex:
+            raise ProtostarGitException(str(ex)) from ex
+
+    return wrapper
+
+
+class Git:
+    @staticmethod
+    def init(path_to_repo: Path):
+        path_to_repo.mkdir(parents=True, exist_ok=True)
+        repo = GitRepository(path_to_repo)
+        repo.init()
+        return repo
+
+    @staticmethod
+    def clone(path_to_repo: Path, repo_to_clone: "GitRepository"):
+        path_to_repo.parent.mkdir(parents=True, exist_ok=True)
+        cloned_repo = GitRepository(path_to_repo)
+        cloned_repo.clone(repo_to_clone.path_to_repo)
+        return cloned_repo
+
+    @staticmethod
+    def from_existing(path_to_repo: Path):
```

Comment:
- `from_*` naming convention is used by static factories. It suggests creating `Git` not `GitRepository`. Maybe `create_from_initialized`
- `path_to_repo` could be provided through the constructor